### ISSUE TITLE
`\int_compare_case:nNnTF` function (and friends)?

### ISSUE THREAD

**Issue by dbitouze** [latex3/latex3#1589]
The `\int_case:nnTF` function:
``` latex
\int_case:nnTF {⟨test int expr⟩}
{
  {⟨int expr case_1⟩} {⟨code case_1⟩}
  {⟨int expr case_2⟩} {⟨code case_2⟩}
  ...
  {⟨int expr case_n⟩} {⟨code case_n⟩}
}
{⟨true code⟩}
{⟨false code⟩}
```

evaluates the `⟨test int expr⟩` and compares its *equality* this in turn to each of the `⟨int expr cases⟩`.

I have a dream: a `\int_compare_case:nNnTF` function which would combine `\int_compare` and `\int_case` that would compare, not only the equality, but more generally the *relation* between them:

``` latex
\int_compare_case:nNnTF {⟨test int expr⟩} ⟨relation⟩
{
  {⟨int expr case_1⟩} {⟨code case_1⟩}
  {⟨int expr case_2⟩} {⟨code case_2⟩}
  ...
  {⟨int expr case_n⟩} {⟨code case_n⟩}
}
{⟨true code⟩}
{⟨false code⟩}
```

This would obviously extend to `\str_`, `\dim_`, `\fp_`, `\str_`, `\file_` and `\pdf_version_`.

(BTW, for the last two, the terminology used in `interface3` for comparison is `⟨comparator⟩` instead of `⟨relation⟩` for the others: is it intentional?)

**Comment by josephwright** [comment]
I'm not sure how that would work - we end up with unpredictable outcomes/code flow.

**Comment by dbitouze** [comment]
If I'm right, the function `\int_compare_case:nnTF {⟨test int expr⟩} ⟨relation⟩ ...` would be, for `⟨relation⟩` being : 

- `=`: just a clone of `\int_case:nnTF {⟨test int expr⟩} ⟨relation⟩ ...`,
- `>=`, `>`, `<=`, `<`: a means of determining the interval in which `⟨test int expr⟩` lies,
- `!=`: a means of determining to which integers `⟨test int expr⟩` differs.

**Comment by muzimuzhi** [comment]
Tip: the `\int_compare_case:nnTF` in the second example of https://github.com/latex3/latex3/issues/1589#issue-2508389635 should have arg-spec `nNnTF`, not `nnTF`.

**Comment by dbitouze** [comment]
If I'm right, the function `\int_compare_case:nnTF {⟨test int expr⟩} ⟨relation⟩ ...` would be, for `⟨relation⟩` being : 

- `=`: just a clone of `\int_case:nnTF {⟨test int expr⟩} ⟨relation⟩ ...`,
- `!=`: a means of determining to which integers `⟨test int expr⟩` differs,
- `>=`, `>` (resp. `<=`, `<`): a means of determining the interval in which `⟨test int expr⟩` lies; of course, in order to avoid unpredictable outcomes/code flow, this would indeed require some precautions, namely `⟨int expr case_1⟩`, `⟨int expr case_2⟩`, ..., `⟨int expr case_n⟩` to be sorted in descendant (resp. ascendant) order.

**Comment by Skillmon** [comment]
`!=` doesn't make much sense, it would be only relevant for really strange border cases since one of the first two comparison values would always match, provided they are distinct. Same is somewhat true for `>=` and `<=` which should be coverable by inverting `<` and `>` (or by just decreasing or increasing the test expression by 1). So I'd argue to only support what `\int_compare:nNnTF` supports if this is added, namely `=`, `>`, and `<`.

**Comment by FrankMittelbach** [comment]
> evaluates the `⟨test int expr⟩` and compares its _equality_ this in turn to each of the `⟨int expr cases⟩`.

The documentation in interface3 is misleading (and should be corrected) it is *not* the case the all cases are tested, the tests stop with the first true value, try
```
\int_case:nnTF{ 3 } {
{ 3 } { first }
{ 4 } { no }
{ 3} {second}
{ 1 } {no} } 
{ true }{false}
```
and this makes lot of sense and speeds up processing a lot of you order the cases so that likely true values come first.

If you want to test for if a given value is within a range or outside a give range the  right thing to do is to make a boolean test with several `\int_if_compare_p...` inside joined with and/or depending on use case. 

I can't see any need for an extra conditional the way you suggest it that isn't already covered by the approach outlined above and the number of situations where you would need something like that are very small, so from my perspective it is a clear no, it only clutters the language as far as I can see.

However, what should be done, is to adjust the documentation to state the the first true result stops the testing and that it is useful to consider ordering the tests based on how likely they become true if that is possible.

**Comment by Skillmon** [comment]
Not sure about directly rejecting it. Sorting into a set of intervals with this could come in handy. One could for instance check whether a single token is a lower case or upper case ASCII letter with this (plus a bit of boiler plate).

The implementation isn't too complicated:

```
\documentclass{article}

\ExplSyntaxOn
\cs_new:Npn \int_case:nNnTF #1
  {
    \exp:w
    \exp_after:wN \__int_case:wNnTF \int_use:N \__int_eval:w #1;
  }
\cs_new:Npn \int_case:nNnT #1#2#3#4
  {
    \exp:w
    \exp_after:wN \__int_case:wNnTF \int_use:N \__int_eval:w #1; #2 {#3} {#4} {}
  }
\cs_new:Npn \int_case:nNnF #1#2#3
  {
    \exp:w
    \exp_after:wN \__int_case:wNnTF \int_use:N \__int_eval:w #1; #2 {#3} {}
  }
\cs_new:Npn \int_case:nNn #1#2#3
  {
    \exp:w
    \exp_after:wN \__int_case:wNnTF \int_use:N \__int_eval:w #1; #2 {#3} {} {}
  }
\cs_new:Npn \__int_case:wNnTF #1; #2
  {
    \token_if_eq_meaning:NNTF > #2
      { \__int_case_aux:nNnnTF { -1 } > }
      {
        \token_if_eq_meaning:NNTF < #2
          { \__int_case_aux:nNnnTF { +1 } < }
          {
            \token_if_eq_meaning:NNF = #2
              { \msg_expandable_error:nnn { int } { unknown-relation } {#2} }
            \__int_case:nnTF
          }
      }
      {#1}
  }
\cs_new:Npn \__int_case_aux:nNnnTF #1#2#3#4#5#6
  {
    \__int_case:nNw {#3} #2 #4 { #3 #1 } {}
      \s__int_mark {#5} \s__int_mark {#6} \s__int_stop
  }
\cs_new:Npn \__int_case:nNw #1#2#3#4
  {
    \int_compare:nNnTF {#1} #2 {#3}
      { \__int_case_end:nw {#4} }
      { \__int_case:nNw {#1} #2 }
  }

\NewDocumentCommand \test { m m }
  {
    \int_case:nNnTF {#1} #2
      {
        {0} {a}
        {3} {b}
        {6} {c}
        {9} {d}
      }
      { TRUE }
      { FALSE }
    \par
  }
\ExplSyntaxOff

\begin{document}
\test{5}>
\test{6}<
\test{-1}>
\test{10}<
\end{document}
```

**Comment by josephwright** [comment]
@FrankMittelbach See my checkin for the docs

**Comment by josephwright** [comment]
@Skillmon  That's not quite the same as the initial request ...

**Comment by Skillmon** [comment]
@josephwright well yes, it is. This is basically what `\int_compare_case:nNnTF` is from the initial request. I just don't support the three unneeded relation types `>=`, `<=`, and `!=` which aren't `N`-type arguments anyway (and additionally see my argumentation above).

**Comment by Skillmon** [comment]
Extending this to `dim` and `fp` is rather easy, because the code would be more or less the same with a few keywords exchanged.

For `str` I'm not sold. Though there is a relational note to `\pdfstrcmp` I really don't see the use case for a relational cases construct.

Same is true for `pdf_version`, I don't see the use case really and think one can simply use `\pdf_version:` or the `_major:`/`_minor:` variants in a cases if it really arises.

And last but not least I don't see a strong use case for a `\file_compare_timestamp`-cases construct.

**Comment by FrankMittelbach** [comment]
@Skillmon that doesn't look like an interface to me let alone an explainable one.  What I could envision is something like
```
\int_case_interval:nnTF  {testvalue}
 {
   {lower1}{upper1}{action1}
   {lower2}{upper2}{action2}
  ...
 }
 {true  case}
 {false case}
```
but not something where you specify a single relation to be used for all testing in all cases.

**Comment by Skillmon** [comment]
@FrankMittelbach why not? It's just a generalisation of `\int_case:nnTF` (in which you implicitly specify the single relation `=` for all cases). Though one could split it into two functions `\int_case_greater:nnTF` and `\int_case_less:nnTF`.

**Comment by Skillmon** [comment]
Or one could go full general and implement

```latex
\int_case_compare:nnTF { testvalue }
  {
    ={case1}{action1}
    >{case2}{action2}
    <{case3}{action3}
  }
  { true }
  { false }
```

**Comment by FrankMittelbach** [comment]
@Skillmon why? because they have not sensible use case that easily reveal what you coding logic is imo. Yes ,they are generalizations I see this, but for what purpose? For example, if you want to sort a value into a set of ranges  (which I think is a valid use case) your full generalization has a hart time doing this if it stops at the first hit, in fact it can't, can it? So what is your real world example why you want the set of tests above?  Perhaps I'm dense, but I don't see it.

**Comment by Skillmon** [comment]
@FrankMittelbach of course the generalisation can parse into a set of intervals, for instance sorting into the intervals [0, 5] and [10, 15] would look like this (admittedly more complicated than the `\int_case_interval:nnTF` function):

```latex
\int_case:nNnF {#1} <
  {
    {0}{false}
    {6}{interval1}
    {10}{false}
    {16}{interval1}
  }
  {false}
```

But I must admit for the single relation generalisation first proposed by me this is the only real use case I can think of, and that would obviously be better covered by `_interval`.

But the other generalisation allowing to specify the relation inside the list of test cases I think is sensible (and can obviously also be used to sort into intervals). So if anything is added I'd argue that one, second choice `interval`, and my first proposition goes to the obscure-ideas corner called bin.

**Comment by Skillmon** [comment]
Also the code for the second generalisation is very concise (the following only defines the `TF` variant, but the others follow easily):

```latex
\cs_new:Npn \int_case_compare:nnTF #1
  {
    \exp:w
    \exp_after:wN \__int_case_compare:wnTF \int_use:N \__int_eval:w #1;
  }
\cs_new:Npn \int_case_compare:nnT #1
\cs_new:Npn \__int_case_compare:wnTF #1;#2#3#4
  {
    \__int_case_compare:nw {#1} #2 ={#1}{}
      \s__int_mark {#3} \s__int_mark {#4} \s__int_stop
  }
\cs_new:Npn \__int_case_compare:nw #1#2#3#4#5
  {
    \int_compare:nNnTF {#1} #2 {#3}
      { \__int_case_end:nw {#4} } % reused internal of `\int_case:nnTF`
      { \__int_case_compare:nw {#1} }
  }
```

**Comment by FrankMittelbach** [comment]
```
\int_case:nNnF {#1} <
  {
    {0}{false}
    {6}{interval1}
    {10}{false}
    {16}{interval2}
  }
  {false}
```

seeing that input syntax to sort into these intervals my initial reaction would be "done by either a mathematicion or a CS person but not by a UI designer" :-)

Can we agree on the fact that this obfucates matters?

From an interface perspective the other one would be ok in my opinion, but from a use case perspective not -- meaning when would you need it and wouldn't in that case a simple boolean test with 2 or 3 `_p` parts not do the trick equally well (and would be understandable to the reader of the code)?

Basically I'm against adding clutter to the language just because you can if there is no use for it ever or nearly ever, and especially not if the language already offers reasonable ways to do it if you need to. For the `case_interval`I can envision reasonable use cases and to me the syntax would be understandable if I see it without remembering the command (or knowing it), but I'm not sure I want to see us rush into providing it just because of that.

**Comment by u-fischer** [comment]
we had a discussion about a range test some time ago: https://github.com/latex3/latex3/issues/1011. 

I too find the suggested syntaxes a bit weird and not easy to parse.  And I don't think that I needed yet such a general test. But I would find a test useful that tests the sign of an integer, so (reusing the syntax above to show what I mean)

~~~~
\int_case_compare:nn { testvalue }
  {
    ={0}{action1}
    >{0}{action2}
    <{0}{action3}
  }
~~~~

**Comment by Skillmon** [comment]
@FrankMittelbach well, we absolutely agree on my first suggestion. My second suggestions is the most versatile, but of course you can build things in a different way (you can also remove `\int_case:nn` from the language and use `\int_compare:nNnT {testvalue} = {case1} { \prg_break:n {action1} }` following that logic though -- yes I know, we don't remove any functions -- or remove all the expandable `case`-functions except for `\bool_case:nnTF` and do the rest via `p`-variants).

`case_interval` could easily become `\bool_case:n` with a small user defined auxiliary of two `\int_compare_p:nNn`.

And I absolutely follow your argumentation of not cluttering the language. Don't understand any of my messages as a vote to actually add this, they are not meant as such. They are more a vote for "if something is added, let it be this one, but don't add it just for the sake of it", I hope that this is clear whenever I provide such things.

**Comment by FrankMittelbach** [comment]
> we had a discussion about a range test some time ago: #1011.
> 
> I too find the suggested syntaxes a bit weird and not easy to parse. And I don't think that I needed yet such a general test. But I would find a test useful that tests the sign of an integer, so (reusing the syntax above to show what I mean)
> 
> ```
> \int_case_compare:nn { testvalue }
>   {
>     ={0}{action1}
>     >{0}{action2}
>     <{0}{action3}
>   }
> ```

If that is considered useful (and we seem to have at least 3 vote for it by now) then something like `\int_compare_relation:nnTF` could perhaps be a name for it as a generalization of `\int_compare:nnTF`. I'm neutral on that.

**Comment by FrankMittelbach** [comment]
> @FrankMittelbach See my checkin for the docs

yes,that looks like a much clearer description. Would it be necessary to say similar things with related conditionals? (there  are essentially copy-paste version of the doc, for example with `\dim_case...` and possibly others.)

**Comment by josephwright** [comment]
> > @FrankMittelbach See my checkin for the docs
> 
> yes,that looks like a much clearer description. Would it be necessary to say similar things with related conditionals? (there are essentially copy-paste version of the doc, for example with `\dim_case...` and possibly others.)

I'll take a look

**Comment by dbitouze** [comment]
@Skillmon I agree with you, `!=` wouldn't make sense.

My initial need is to perform different actions (defining the width of the spine of a cover book) depending on the interval (among a predefined set of intervals) in which a given integer (the total number of pages in the book) lies. But this may be a niche use case.

**Comment by dbitouze** [comment]
> I think this is done in the docs

Sorry but what is done?

**Comment by josephwright** [comment]
@dbitouze Many comments ago, @FrankMittelbach said

> However, what should be done, is to adjust the documentation to state the the first true result stops the testing and that it is useful to consider ordering the tests based on how likely they become true if that is possible.

which is what I have done - I'm wondering if there is any more to do as the overall plan seemed at best to be quite niche (at least to me).
